src/context/distributor.hpp

```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include "cocaine/memory.hpp"
+#include "cocaine/dynamic.hpp"
+#include "cocaine/errors.hpp"
+
+namespace cocaine {
+
+// distributor of elements of Pool
+// Pool should provide value_type typedef and 'operator[](size_t)' call
+// value_type should provide operator* and oprator->,
+// resulting type should provide 'double utilization()' call with returned values in [0.0 1.0] range
+template <class Pool>
+struct distributor {
+    using result_type = typename Pool::value_type;
+
+    virtual
+    ~distributor() {}
+
+    virtual
+    auto
+    next(Pool& pool) -> result_type& = 0;
+};
+
+
+// Distributes elements of the Pool according only to 'utilization()' of the elements.
+// Simply returns the lowest value.
+template <class Pool>
+struct load_distributor: public distributor<Pool> {
+    using result_type = typename Pool::value_type;
+
+    load_distributor() {}
+
+    auto
+    next(Pool& pool) -> result_type& override {
+        auto comp = [](const result_type& lhs, const result_type& rhs) {
+            return lhs->utilization() < rhs->utilization();
+        };
+        return *std::min_element(pool.begin(), pool.end(), comp);
```

Comment:
If pool is (unexpectedly) empty it would be reference to pool::end here.

Reply:
Pool can not be empty afaik